### PR TITLE
Convert PassManager.passes to property

### LINKED ISSUE #8239: PassManager passes should be a property
### What should we add?

`PassManager.passes` should be defined as a `@property` instead of as a callable/method.

### PR DESCRIPTION
**Pull request by pedrorrivero**
### Summary
Convert `PassManager.passes` to property and introduce deprecation warning when called

Breaking change
Closes #8239 
Changelog: New Feature
Changelog: Deprecation

### CODE REVIEW

File: qiskit/transpiler/passmanager.py [hunk 1]
```diff
@@ -320,6 +322,14 @@ def passes(self) -> List[Dict[str, BasePass]]:
...
                 item["flow_controllers"] = {}
             ret.append(item)
+
+        def call(self):
+            warn("The `PassManager.passes()` method is deprecated as of 0.22.0, and "
+                  "will be removed no earlier than 3 months after that "
+                  "release date. You should use the `PassManager.passes` property instead.", DeprecationWarning, stacklevel=2)
+            return self
+
+        ret.__call__ = call
```
Comment (mtreinish) [inline qiskit/transpiler/passmanager.py:332]: I think for this to work in all situations you'll have to create a `list` subclass (or a `MutableSequence` type) and return that. Monkey patching a call definition on doesn't get preserved in a parallel context which is the source of the failures in the integration tests: https://github.com/Qiskit/qiskit-terra/runs/7046189043?check_suite_focus=true

**Review by mtreinish — COMMENTED** [review]
Thanks for doing this, I agree with the direction, the `passes` attribute should have always been a property. I left some inline comments. Also this probably could use some tests and a release note  documenting the new access pattern.

**Review by jakelishman — COMMENTED** [review]
I don't care about / am not responsible for these particular objects, so please don't take this as a decision, but I'm opposed to making this sort of object a `property` in general.  I've seen these sorts of properties cause a lot of problems elsewhere in Qiskit for two main reasons:
- it does non-trivial work for something which, at the call-site looks like a free action.  This leads to inefficient code if `pm.passes` is accessed in a loop, with no obvious reason.,
- returning a mutable structure from an attribute makes it seem like modifying the structure will affect the underlying object, e.g. `pm.passes.append(new_pass)`, but here it won't actually have any effect.

Because of these, to me `passes` is not logically a static property of this object, and so it's right that it needs a function-call to create.

(Also, since the interface is already a function and requires a deprecation to change, we need a very good reason to change it, and there's insufficient reason here.)

**Comment by mtreinish** [comment]
That's a fair point, I didn't look too closely at the implementation I assumed the passes list was just a static return of the internal `_pass_sets` attribute but looking at the implementation in more detail @jakelishman is correct that this is doing restructuring of the output which is dynamic and done iterating over the contents and not just the equivalent of getattr of the private attribute. I agree that we shouldn't be doing work like this in a `property`. Why it's doing this transform I'm not sure tbh I don't really see any advantage to how it's tweaking things.

**Comment by jakelishman** [comment]
Given the above comment, and the same discussion of merge conflicts / changes as in #8277, I'll close this as stale / possibly not needed, but feel free to re-open or open a new PR if there's more to discuss.

**Pull request closed without merging**